[openblas] Make build_lapack=True default for >= 0.3.21

samuel-emrys opened the pull request:
Version 0.3.21 of openblas introduced the ability to build lapack from C. Make this default to enable CCI recipes to use lapack when depending on openblas.

Additionally, fix an inconsistency in version constraint. Review was provided in https://github.com/conan-io/conan-center-index/pull/19808 to constrain it to 0.3.24 as this was the latest version available in CCI. However, functionality to build lapack was added in 0.3.21, so it doesn't really make sense to exclude versions between 0.3.21 and 0.3.24 from building lapack should they be added at a later date.

Closes #7361

Specify library name and version:  **openblas/0.3.24**

---

- [x] I've read the [contributing guidelines](https://github.com/conan-io/conan-center-index/blob/master/CONTRIBUTING.md).
- [x] I've used a [recent](https://github.com/conan-io/conan/releases/latest) Conan client version close to the [currently deployed](https://github.com/conan-io/conan-center-index/blob/master/.c3i/config_v1.yml#L6).
- [x] I've tried at least one configuration locally with the [conan-center hook](https://github.com/conan-io/hooks.git) activated.

samuel-emrys commented:
@uilianries any idea what's going on with CI here?

samuel-emrys commented:
> @samuel-emrys Could you please, show a link or document proving that support, just in case?

This PR only makes the lapack build default. Support for building lapack from C was [approved by yourself](https://github.com/conan-io/conan-center-index/pull/19808#pullrequestreview-1710109471) in #19808:

> Okay, it seems be following https://github.com/OpenMathLib/OpenBLAS/releases/tag/v0.3.21

From the v0.3.21 release notes:

> when no Fortran compiler is available, OpenBLAS builds will now automatically
build LAPACK from an f2c-converted copy of LAPACK 3.9.0 unless the NO_LAPACK option
is specified (more recent releases make too heavy use of Fortran90+ features to be easily convertible to C)

uilianries reviewed (changes requested):
I'm not confident accepting this PR now, because there is no fortran support when building this package in CCI:

https://c3i.jfrog.io/c3i/misc/logs/pr/20894/3-linux-gcc/openblas/0.3.24//729e2486182de25560f46a58cc644fb2e1eb8b2f-build.txt

The version 0.3.24 will build with lapack, but no fortran is available, so the generated package is not good, someone will open an issue complaining about this lack of support for sure.

I'll check with Conan team to include fortran to Linux docker images, at least, we could solve it partially.

samuel-emrys commented:
> I'm not confident accepting this PR now, because there is no fortran support when building this package in CCI:
> 
> https://c3i.jfrog.io/c3i/misc/logs/pr/20894/3-linux-gcc/openblas/0.3.24//729e2486182de25560f46a58cc644fb2e1eb8b2f-build.txt
> 
> The version 0.3.24 will build with lapack, but no fortran is available, so the generated package is not good, someone will open an issue complaining about this lack of support for sure.


@uilianries I'm not sure I follow what you mean. What are you drawing attention to in those logs? All that related to fortran that I could find was this:

> openblas/0.3.24: Building LAPACK without Fortran compiler

which is the point of this PR. We can remove any dependency on fortran by building lapack from C sources. I'm not sure why anybody would complain, as it's reasonably easy to trigger a fortran build by defining the `fortran` key in `compiler_executables` if desired as per https://github.com/conan-io/conan-center-index/blob/e35433047da05a6210a7daf2bd232210d431f70c/recipes/openblas/all/conanfile.py#L40-L44

> I'll check with Conan team to include fortran to Linux docker images, at least, we could solve it partially.

#19808 removed the dependence on fortran so that we can build lapack with a C++ only toolchain - this seems to be the most desirable solution to me - what am I missing? `-lgfortran` is only added when the user specifies a value for `compiler_executables={"fortran": "/path/to/gfortran"}`, which would not be default for openblas >= 0.3.21 since this is not included in the conan profiles.

Lack of lapack support on CCI due to lack of a fortran compiler has been an ongoing issue for years now (https://github.com/conan-io/conan-center-index/pull/5260, https://github.com/conan-io/conan-center-index/pull/15556). I would really rather solve this with the simplest solution we have which is to build it from C. It's not perfect - it version locks lapack to 3.9.0, but it gives CCI minimum viability uniformly across all operating systems. Other math packages only provide limited utility without lapack, such as armadillo.

If we can explore adding fortran compilers to conan CI infrastructure to add a proper lapack recipe or build it here, that would be good as well, but I strongly suggest we should go for the quick win here.

samuel-emrys commented:
@uilianries can you engage on this more/elaborate on your reasoning? it doesn't make a lot of sense to me and having lapack available is a critical requirement for math libraries in CCI.

cc @RubenRBS @memsharded

uilianries commented:
@samuel-emrys indeed is a problem also related to the CI. I'll take a look to provide fortran via Docker images, so it should be able to build anything. Some images are prepared already (conanio/gcc11-ubuntu16.04), but Clang images need and extra build (checking to use llvm flang). For legacy images will be more complicated, because some of them are using ubuntu-toolchain-r or llvm repository and fortran is not always available.

samuel-emrys commented:
Adding fortran support to conan ci's is a good idea and is something that should be progressed independently of this PR, but it does open a can of worms with respect to binary compatibility - conan doesn't currently have a way to describe the ABI or runtime requirements that building a package with ifort or gfortran implies.

Building the fortran components from sources that have been converted to C really is the best way forward here in my opinion, because none of those ABI considerations will factor in here - it becomes a plain C library.

This is what we want as default wherever possible until conan has the features to support polyglot considerations.

samuel-emrys commented:
@uilianries @RubenRBS what's the best way to have a conversation about this and or an alternate pathway to getting lapack/other fortran libs on cci? Here or slack? I don't understand the rationale for not accepting this - this seems to be the best solution for conan because it bypasses the fortran requirement entirely, and this has been an ongoing issue for **years** (https://github.com/conan-io/conan-center-index/pull/5260, https://github.com/conan-io/conan-center-index/pull/15556) - getting a resolution here would be good and would enable a number of scientific computing libraries to work unimpeded.

// cc @memsharded

samuel-emrys commented:
Just to be clear in case it was missed above, the reason I'm confused is that from the [release notes](https://github.com/OpenMathLib/OpenBLAS/releases/tag/v0.3.21):

> when no Fortran compiler is available, OpenBLAS builds will now automatically
build LAPACK from an f2c-converted copy of LAPACK 3.9.0 unless the NO_LAPACK option
is specified (more recent releases make too heavy use of Fortran90+ features to be easily convertible to C)

This means that the sources have been converted from fortran to c, so this change triggers a build from the ***C sources*** meaning that there's no dependency on fortran at all here - this is essentially a C library, which sidesteps the problem of using fortran at all. I would imagine is the easiest path forward for conan-center-index.

uilianries commented on recipes/openblas/all/conanfile.py:
```diff
@@ -46,6 +46,8 @@ def _fortran_compiler(self):
     def config_options(self):
         if self.settings.os == "Windows":
             del self.options.fPIC
+        if Version(self.version) >= "0.3.21":
```
```suggestion
        # INFO: When no Fortran compiler is available, OpenBLAS builds LAPACK from an f2c-converted copy of LAPACK unless the NO_LAPACK option is specified
```

I would include a note, just to be clear about it.

samuel-emrys commented:
These failures are as a result of https://github.com/conan-io/conan-center-index/commit/d34e4a1870cd0790a3ee791936ab7fee7d24f0ed, which introduces a native Macos armv8 runner to the c3i build matrix. armv8 builds of openblas have to date, only been explicitly disallowed for cross builds, and so an "invalid" configuration has made it through here and is causing failures.

I've attempted to add armv8 compatibility in this PR (by hotfixing https://github.com/OpenMathLib/OpenBLAS/pull/4142
), but if this doesn't work, I'll exclude native armv8 builds from the valid build matrix and leave the improvement for a future PR.

This also relates to https://github.com/conan-io/conan-center-index/pull/21485

samuel-emrys commented:
I've reverted the changes to add armv8 compatibility as they weren't immediately working (and they're beyond the scope of this PR), and I've added an additional exclusion for `armv8` to the `validate()` method. This can be followed up in https://github.com/conan-io/conan-center-index/pull/21485 which aims to add this compatibility.

uilianries reviewed (approved):
LGTM
